### PR TITLE
Expose build information in dynamic lib

### PR DESCRIPTION
**Pull request by guoyuhong**
### Description

1. Add Build Info API to onnx.
2. Fix compile error while building onnxruntime_benchmark in MacOs.

### Motivation and Context

1. When Onnxruntime lib is serving online, we need a way to detect how this lib is built. This PR helps the developer to get the build information using `strings` such as git branch, git commit id, build type and cmake cxx flags, which is showed as follows.

image

image

If the build env has no git, there will be no git related infor:

image

3. Fix the following compile error while building benchmark in MacOs.
image

### CODE REVIEW

**Comment by guoyuhong** [comment]
@microsoft-github-policy-service agree

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -1229,6 +1229,15 @@ if (onnxruntime_USE_VITISAI)
   endif()
 endif()
 
+execute_process(COMMAND git log -1 --format=%h
```
Comment (snnn) [inline cmake/CMakeLists.txt:1232]: Some of our build environments do not have git. So this would not work.

**Review by snnn — CHANGES_REQUESTED** [review]
No git.

File: onnxruntime/core/framework/tuning_context_impl.h [hunk 2]
```diff
@@ -215,13 +215,11 @@ Status TuningResultsValidator::ValidateOrtVersion(const std::string& value) cons
...
 std::string TuningResultsValidator::GetOrtGitCommit() const {
-  // TODO:
-  return "";
+  return ORT_GIT_COMMIT;
 }
 
 Status TuningResultsValidator::ValidateOrtGitCommit(const std::string& value) const {
-  // TODO:
-  ORT_UNUSED_PARAMETER(value);
+  ORT_RETURN_IF(value != ORT_GIT_COMMIT, "onnxruntime git commit mismatch");
```
Comment (guoyuhong) [inline onnxruntime/core/framework/tuning_context_impl.h:222]: @snnn As you said there could be build envs that have no git, is it better to revert this change for `ValidateOrtGitCommit `?

File: cmake/CMakeLists.txt (hunk 1 above)
Comment (guoyuhong) [reply cmake/CMakeLists.txt:1233]: @snnn It does not matter whether the build envs have git or not. The cmake command won't fail. Instead, it will output to the variable `ORT_GIT_COMMIT` & `ORT_GIT_BRANCH` with an empty string. To make this clearer, I added a comment here and explicitly change the empty string to "Not-Available". An example is show below:

image

Comment (guoyuhong) [reply cmake/CMakeLists.txt]: @snnn I have reverted the change related to git in `cmake/onnxruntime_config.h.in`. For `GetBuildInfoString` in envs that has no git, it will returns the build type and cxx flags as follows. If the env has git, it will contain more information. Is it better?
I have also reverted the change in `tuning_context_impl.h` since the commit id requires git.
image

File: onnxruntime/core/framework/tuning_context_impl.h (hunk 2 above)
Comment (guoyuhong) [reply onnxruntime/core/framework/tuning_context_impl.h]: Never mind, reverted.

File: include/onnxruntime/core/session/onnxruntime_cxx_api.h [hunk 3]
```diff
@@ -127,6 +127,13 @@ inline const OrtApi& GetApi() noexcept { return *Global<void>::api_; }
 /// <returns>version string major.minor.rev</returns>
 std::string GetVersionString();
 
+/// <summary>
+/// This function returns the onnxruntime build information: including git branch,
+/// git commit id, build type(Debug/Release/RelWithDebInfo) and cmake cpp flags.
+/// </summary>
+/// <returns>string</returns>
+std::string GetBuildInfoString();
```
Comment (xadupre) [inline include/onnxruntime/core/session/onnxruntime_cxx_api.h:135]: It would be helpful to make it available in the python bindings as well. Any test to check its value?

Comment (guoyuhong) [reply include/onnxruntime/core/session/onnxruntime_cxx_api.h:135]: That's a good point. I will take a look at how to add the python binding and the test to cover.

Comment (guoyuhong) [reply include/onnxruntime/core/session/onnxruntime_cxx_api.h:135]: @xadupre I have added the build info to python bindings. The figure below shows the effect by some temporary code.
image

**Comment by snnn** [comment]
+ @pranavsharma

All API changes need to be reviewed by him.

**Review by pranavsharma — DISMISSED** [review]

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 4]
```diff
@@ -619,7 +619,8 @@ struct OrtApiBase {
    *   older than the version created with this header file.
    */
   const OrtApi*(ORT_API_CALL* GetApi)(uint32_t version)NO_EXCEPTION;
-  const char*(ORT_API_CALL* GetVersionString)(void)NO_EXCEPTION;  ///< Returns a null terminated string of the version of the Onnxruntime library (eg: "1.8.1")
+  const char*(ORT_API_CALL* GetVersionString)(void)NO_EXCEPTION;    ///< Returns a null terminated string of the version of the Onnxruntime library (eg: "1.8.1")
```
Comment (snnn) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:622]: In general we should use const ORTCHAR_T* instead of const char*, because on Windows usually we prefer to use wide char.

File: cmake/CMakeLists.txt [hunk 5]
```diff
@@ -1229,6 +1229,19 @@ if (onnxruntime_USE_VITISAI)
   endif()
 endif()
 
+set(ORT_BUILD_INFO "ORT Build Info: ")
+find_package(Git)
```
[the remark this answers is not in the document]
Comment (snnn) [reply cmake/CMakeLists.txt:1233]: If from the beginning we were very strict on this, we won't see such an issue: https://github.com/microsoft/onnxruntime/issues/15678

**Comment by snnn** [comment]
LGTM. 

I hope the "const char*" could be changed to "const ORTCHAR_T*" to avoid confusion.

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 4 above)
Comment (guoyuhong) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:622]: Good point. `GetVersionString` is originally defined using `char *`. I will change both.

File: cmake/CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (guoyuhong) [reply cmake/CMakeLists.txt:1233]: > If from the beginning we were very strict on this, we won't see such an issue: #15678
By this, you mean the `ORTCHAR_T*` problem, right? Yes, I have changed `const char *` to `const ORTCHAR_T *`

**Comment by linkerzhang** [comment]
Ahahaha, a good try to communicae with Microsoft ORT team. Antgroup will try to contribute more back to ORT based on what we did during last years.  ORT rocks!

File: onnxruntime/core/session/onnxruntime_c_api.cc [hunk 6]
```diff
@@ -2762,10 +2763,14 @@ ORT_API(const OrtApi*, OrtApis::GetApi, uint32_t version) {
   return nullptr;  // Unsupported version
 }
 
-ORT_API(const char*, OrtApis::GetVersionString) {
+ORT_API(const ORTCHAR_T*, OrtApis::GetVersionString) {
   return ORT_VERSION;
 }
 
+ORT_API(const ORTCHAR_T*, OrtApis::GetBuildInfoString) {
+  return ORT_BUILD_INFO;
```
Comment (snnn) [inline onnxruntime/core/session/onnxruntime_c_api.cc:2771]:
```
return ORT_TSTR(ORT_VERSION);
```

Comment (snnn) [reply onnxruntime/core/session/onnxruntime_c_api.cc:2771]: return ORT_BUILD_INFO;

File: onnxruntime/core/session/onnxruntime_c_api.cc [hunk 7]
```diff
@@ -2762,8 +2763,12 @@ ORT_API(const OrtApi*, OrtApis::GetApi, uint32_t version) {
   return nullptr;  // Unsupported version
 }
 
-ORT_API(const char*, OrtApis::GetVersionString) {
-  return ORT_VERSION;
+ORT_API(const ORTCHAR_T*, OrtApis::GetVersionString) {
+  return ORT_TSTR(ORT_VERSION);
```
Comment (snnn) [inline onnxruntime/core/session/onnxruntime_c_api.cc:2767]: I probably gave you a bad suggestion. It did not work.

Comment (guoyuhong) [reply onnxruntime/core/session/onnxruntime_c_api.cc]: @snnn Yes, it looks like that `ORT_TSTR` should not be applied to a macro, which may lead to undefined variables. I define another macro `ORT_TSTR_ON_MACRO` to do so. I create a code snippet to replay the compile error in Linux and found the new macro could solve the problem. I hope it will also work in Windows which I don't have.
image

File: java/src/main/native/ai_onnxruntime_OnnxRuntime.c [hunk 8]
```diff
@@ -76,8 +76,13 @@ JNIEXPORT jobjectArray JNICALL Java_ai_onnxruntime_OnnxRuntime_getAvailableProvi
 JNIEXPORT jstring JNICALL Java_ai_onnxruntime_OnnxRuntime_initialiseVersion
   (JNIEnv * jniEnv, jclass clazz) {
   (void)clazz;  // required JNI parameter not needed by functions which don't access their host class.
-  const char* version = OrtGetApiBase()->GetVersionString();
+  const ORTCHAR_T* version = OrtGetApiBase()->GetVersionString();
   assert(version != NULL);
+#ifdef _WIN32
+  size_t len = wcslen(version);
+  jstring versionStr = (*jniEnv)->NewString(jniEnv, (const jchar*)version, len);
```
Comment (snnn) [inline java/src/main/native/ai_onnxruntime_OnnxRuntime.c:83]: If I remember correctly, Java use UTF-16-BE while most Windows use UTF-16-LE.  @Craigacp, is it true?

Comment (Craigacp) [reply java/src/main/native/ai_onnxruntime_OnnxRuntime.c:83]: We don't transform the output of jstring->GetStringChars to change UTF-16 endianness on Windows, and it can load & save things to those paths correctly. So I would expect it to be the same on platforms that we currently test in the CI. If it's the same on all platforms is beyond my knowledge of Windows. It should be straightforward to modify https://github.com/microsoft/onnxruntime/blob/main/java/src/test/java/ai/onnxruntime/InferenceTest.java#L74 to test that the output is correct. I haven't modified that test to check the string because it'll need updating every time the version number is bumped so it seems like it would fail a bunch of pipelines.

**Review by snnn — APPROVED** [review]

**Comment by guoyuhong** [comment]
Thanks to @snnn @HectorSVC for your patience. :-)

**Comment by snnn** [comment]
Happy holidays!

**Comment by yuslepukhin** [comment]
What is the reason why simple version information is being returned as ORTTCHAR instead of UTF-8 as all other strings inside ORT? That typedef was introduced for paths, otherwise all strings are expected to contain UTF-8. It is just makes coding so much more complicated for what essentially is a stupid version string.

**Comment by guoyuhong** [comment]
@yuslepukhin Sorry for the inconvenience. I just followed the comment from @snnn , and not expected that would bring so much trouble. I learned that "We define all strings as UTF-8 internally, except Windows paths.". Thanks.
